test/FunctionalTests/Microsoft.OData.Core.Tests/JsonLight/ODataJsonLightWriterTests.cs

```diff
@@ -738,6 +745,114 @@ public async Task WriteEntityReferenceLinkAsync()
                 result);
         }
 
+        /// <summary>
+        /// Gets the name of the caller method of this method
+        /// </summary>
+        /// <param name="caller">The string that the method name of the caller will be written into</param>
+        /// <returns>The name of the caller method of this method</returns>
+        public static string GetCurrentMethodName([System.Runtime.CompilerServices.CallerMemberName] string caller = null)
+        {
+            return caller;
+        }
+
+#if !NETCOREAPP1_1
+        /// <summary>
+        /// Writes a resource as the response to a request where the URL ends with a combined cast and key segment
+        /// </summary>
+        /// <returns><see cref="void"/></returns>
+        [Fact]
+        public static async Task WriteContextWithDerivedTypeCastAndKeySegmentAsync()
+        {
+            var domain = new Uri("http://tempuri.org");
+            var requestUrl = new Uri(domain, "/orders('1')/products/ns.derivedProduct('2')");
+            var serviceSideResponseResource = new ODataResource
+            {
+                TypeName = "ns.product",
+                Properties = new List<ODataProperty>
+                {
+                    new ODataProperty
+                    {
+                        Name = "id",
+                        Value = "1",
+                        SerializationInfo = new ODataPropertySerializationInfo
+                        {
+                            PropertyKind = ODataPropertyKind.Key
+                        },
+                    },
+                    new ODataProperty
+                    {
+                        Name = "name",
+                        Value = "somename",
+                    },
+                },
+            };
+            var expectedResponsePayload = 
+                "{" +
+                    "\"@odata.context\":\"http://tempuri.org/$metadata#orders('1')/products/$entity\"," +
+                    "\"id\":\"1\"," +
+                    "\"name\":\"somename\"" + 
+                "}";
+
+            // load the CSDL from the embedded resources
+            var assembly = Assembly.GetExecutingAssembly();
+            var currentMethod = GetCurrentMethodName();
+            var csdlResourceName = assembly.GetManifestResourceNames().Where(name => name.EndsWith($"{currentMethod}.xml")).Single();
+
+            // parse the CSDL
+            IEdmModel model;
+            using (var csdlResourceStream = assembly.GetManifestResourceStream(csdlResourceName))

```

Comment:
Loading from manifest Resource looks good. 
But, If it's only used in one test case, is it too heavy? :)

Reply:
I think it gives a straightfoward way to have a different model for different tests. I think it would be good to start doing this in other tests as well going forward. I'd like to keep it this way unless you have a strong objection.

Reply:
Because I've started using this for new issues as well to good effect, I'm going to keep it this way. We can refactor in the future if we need to.